account-kit/smart-contracts/src/light-account/accounts/viem-base.ts

```diff
@@ -0,0 +1,344 @@
+  TLightAccountType extends LightAccountType,
+  TLightAccountVersion extends
+    LightAccountVersion<TLightAccountType> = LightAccountVersion<TLightAccountType>,
+  TTransport extends Transport = Transport,
+  TSigner extends SmartAccountSigner = SmartAccountSigner,
+> = {
+  client: PublicClient<TTransport, Chain>;
+  signer: TSigner;
+  abi: Abi;
+  accountAddress: Address;
+  type: TLightAccountType;
+  version: TLightAccountVersion;
+  getFactoryArgs: () => Promise<{
+    factory?: Address | undefined;
+    factoryData?: Hex | undefined;
+  }>;
+};
+
+export async function createViemLightAccountBase<
+  TLightAccountType extends LightAccountType,
+  TLightAccountVersion extends
+    LightAccountVersion<TLightAccountType> = LightAccountVersion<TLightAccountType>,
+  TTransport extends Transport = Transport,
+  TSigner extends SmartAccountSigner = SmartAccountSigner,
+>({
+  client,
+  signer,
+  abi,
+  version,
+  type,
+  accountAddress,
+  getFactoryArgs,
+}: CreateViemLightAccountBaseParams<
+  TLightAccountType,
+  TLightAccountVersion,
+  TTransport,
+  TSigner
+>): Promise<
+  ViemLightAccountBase<TSigner, TLightAccountType, TLightAccountVersion>
+> {
+  const encodeUpgradeToAndCall = async ({
+    upgradeToAddress,
+    upgradeToInitData,
+  }: {
+    upgradeToAddress: Address;
+    upgradeToInitData: Hex;
+  }): Promise<Hex> => {
+    const storage = await getStorageAt(client, {
+      address: accountAddress,
+      slot: "0x360894a13ba1a3210667c828492db98dca3e2076cc3735a920a3ca505d382bbc",
+    });
+
+    if (storage == null) {
+      throw new Error(
+        "Failed to get storage slot: 0x360894a13ba1a3210667c828492db98dca3e2076cc3735a920a3ca505d382bbc",
+      );
+    }
+
+    const implementationAddresses = Object.values(
+      AccountVersionRegistry[type],
+    ).map(
+      (x) =>
+        x.addresses.overrides?.[client.chain.id]?.impl ??
+        x.addresses.default.impl,
+    );
+
+    if (
+      fromHex(storage, "number") !== 0 &&
+      !implementationAddresses.some((x) => x === trim(storage))
+    ) {
+      throw new Error(
+        `could not determine if smart account implementation is ${type} ${String(
+          version,
+        )}`,
+      );
+    }
+
+    return encodeFunctionData({
+      abi,
+      functionName: "upgradeToAndCall",
+      args: [upgradeToAddress, upgradeToInitData],
+    });
+  };
+
+  const get1271Wrapper = (
+    hashedMessage: Hex,
+    version: string,
+  ): TypedDataDefinition => {
+    return {
+      domain: {
+        chainId: client.chain.id,
+        name: type,
+        verifyingContract: accountAddress,
+        version,
+      },
+      types: {
+        LightAccountMessage: [{ name: "message", type: "bytes" }],
+      },
+      message: {
+        message: hashedMessage,
+      },
+      primaryType: "LightAccountMessage",
+    };
+  };
+
+  const prepareSignature = async (
+    message: string | TypedDataDefinition,
+    isTypedData: boolean = false,
```

Comment:
❓ do we really need this `isTypedData` or can we just infer it based on the `message`?

also does this need to support a message like `{raw: Hex}` too or nah?

Reply:
we might be able to, but I kinda like "explicit over implicit"